c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -223,7 +231,325 @@ struct Block {
+even multiples of that size. Many deep learning models follow this behavior.
+However, one common exception is when the batch size changes slightly from one
+iteration to the next, e.g. in batched inference. When the program runs
+initially with batch size N, it will make allocations appropriate for that size.
+If in the future, it runs at size N - 1, the existing allocations will still be
+big enough. However, if it runs at size N + 1, then it will have to make new
+allocations that are slightly larger. Not all the tensors are the same size.
+Some might be (N + 1)*A and others (N + 1)*A*B where A and B are some non-batch
+dimensions in the model. Because the allocator reuses existing allocations when
+they are big enough, some number of (N + 1)*A allocations will actually fit in
+the already existing N*B*A segments, though not perfectly. As the model runs it
+will partially fill up all of these segments leaving unusable free slices of
+memory at the end of these segments. The allocator at some point will need to
+cudaMalloc a new (N + 1)*A*B segment. If there is not enough memory, there is
+now no way to recover the slices of memory that are free at the end of existing
+segments. With models 50+ layers deep, this pattern might repeat 50+ times
+created many slivers.
+
+Approach
+
+Expandable segments allows the allocator to create a segment initially and then
+expand its size later when more memory is needed. Instead of making one segment
+per allocation, it tries to make one segment (per stream) that grows as
+necessary. Now when the N + 1 case runs, the allocations will tile nicely into
+the one large segment until it fills up. Then more memory is requested and
+appended to the end of the segment. This process does not create as many slivers
+of unusable memory, so it is more likely to succeed at finding this memory.
+
+Implementation
+
+The expandable_segments:True option is used to enable/disable this behavior. We
+use cuda's low-level memory APIs, which are similar to mmap, to extend the
+memory segments. These APIs separate the allocation of physical memory
+(cuMemCreate) from the allocation of virtual address space (cuMemAddressReserve)
+and the associate between them cuMemMap/cuMemSetAccess.
+
+When we allocate a new segment, we allocate enough address space to map
+basically the entire physical memory of the GPU (there is 256TiB of address
+space), but we only map enough physical memory to handle the current amount of
+memory needed by the program. As more is requested, we add more physical memory
+to the segment. This can work at the granularity of GPU pages which are 2MiB
+currently.
+
+If we end up out of memory, we can unmap all the memory in our segmenet
+corresponding to empty physical pages, and return it to CUDA for use at another
+address in the segment or in a segment for a different stream.
+
+A current limitation of CUDA's API is that physical memory cannot be split up
```

Comment:
When you described this to me previously, you said that the mmap-ing requires a whole device sync. Worth putting that in the speed description here ?